crates/bevy_ui/src/render/mod.rs

```diff
@@ -398,20 +402,31 @@ pub fn prepare_uinodes(
 
     let mut start = 0;
     let mut end = 0;
-    let mut current_batch_handle = Default::default();
+    let mut stored_batch_handle = DEFAULT_IMAGE_HANDLE.typed();
     let mut last_z = 0.0;
+    let default_id = DEFAULT_IMAGE_HANDLE.id();
+
     for extracted_uinode in &extracted_uinodes.uinodes {
-        if current_batch_handle != extracted_uinode.image {
-            if start != end {
-                commands.spawn(UiBatch {
-                    range: start..end,
-                    image: current_batch_handle,
-                    z: last_z,
-                });
-                start = end;
+        let mode = if extracted_uinode.image.id() != default_id {
+            if stored_batch_handle.id() != default_id {
+                if stored_batch_handle.id() != extracted_uinode.image.id() {
+                    if start != end {
+                        commands.spawn(UiBatch {
+                            range: start..end,
+                            image: stored_batch_handle,
+                            z: last_z,
+                        });
+                        start = end;
+                    }
+                    stored_batch_handle = extracted_uinode.image.clone_weak();
+                }
+            } else {
+                stored_batch_handle = extracted_uinode.image.clone_weak();
```

Comment:
AbleHeathen on discord mentioned another possible improvement:

```rust
if is_batch_default || !is_batch_extracted {
    stored_batch_handle = extracted_uinode.image.clone_weak();
}

if !is_batch_extracted && start != end {
    commands.spawn(UiBatch { 
      range: start..end,
      image: stored_batch_handle,
      z: last_z,
    });
    start = end;
}
```

Didn't check the logic, but seems even better.

Reply:
The conditionals are all necessary I think. It's more complicated than it looks at first when there are multiple different textures being queued. 

This is the result we want:
![prepare_uinodes](https://github.com/bevyengine/bevy/assets/27962798/a6027403-1cc9-49fc-91b1-8b5bbe9a5e19)

But with your changes we get:
![prepare_uinodes_2](https://github.com/bevyengine/bevy/assets/27962798/b6a4ad7f-8cc7-4289-9c08-c7bf9cc6e0c8)

You are right that it can be cleaned up though, I'll see what I can do.

Reply:
I triple checked my code. It wasn't easy to do (hence the remark about it being difficult to reason about) but I've done it slowly and step by step. I'm not 100% sure, but I'm fairly confident the logic is equivalent, at least for the following code: 

```rust
    let mode = if is_extracted_default { UNTEXTURED_QUAD } else { TEXTURED_QUAD };

    if !is_extracted_default {
        if is_batch_default {
            stored_batch_handle = extracted_uinode.image.clone_weak();
        } else if !is_batch_extracted {
            stored_batch_handle = extracted_uinode.image.clone_weak();
            if start != end {
                commands.spawn(UiBatch {
                    range: start..end,
                    image: stored_batch_handle,
                    z: last_z,
                });
                start = end;
            }
        }
    }
```

I made a gist so the changes can be visualized: <https://gist.github.com/nicopap/2511c8761eccb059757c88e4d6d69689/revisions>

Reply:
No this is definitely incorrect. 
`stored_batch_handle` holds the handle of the texture for the next `UiBatch`.
Then after that `UiBatch` is spawned, we set `stored_batch_handle` to the texture for the next batch from `extracted_uinode`.

Your change overwrites `stored_batch_handle` first before the `UiBatch` is created so the batch will be spawned with the wrong texture. Also, it doesn't look like it will compile because `UiBatch` takes ownership of `stored_batch_handle`, so the value can't be used in the next iteration of the loop.

Reply:
I think the lastest version I just pushed is looking ok now. You were right that `stored_batch_handle` could be set at the end of the code paths, instead of in both of the different if blocks.

Reply:
Agreed that this section of code is still hard to understand. I don't feel confident reviewing it atm.